### PR TITLE
Fet/rest relation endpoints

### PR DESCRIPTION
**Pull request by mnelson4**
## Problem this Pull Request solves
https://github.com/eventespresso/event-espresso-core/issues/790
Adds relation endpoints for simplifying adding and removing relations between individual model objects.
Also, fixes a bug/feature in the models that facilitated multiple entries in join tables having the same foreign keys.
Please see documentation added in this PR for more: https://github.com/eventespresso/event-espresso-core/blob/FET/rest-relation-endpoints/docs/C--REST-API/ee4-rest-api-editing-relations.md

## How has this been tested

New features unit tested, so we should just verify there are no regressions by...
* [ ] Verify the mobile apps still work fine
* [x] Verify the attendee mover works fine

## Checklist

* [ ] I have read the documentation relating to systems affected by this pull request, see https://github.com/eventespresso/event-espresso-core/tree/master/docs
* [ ] User input is adequately validated and sanitized
* [ ] all publicly displayed strings are internationalized (usually using `esc_html__()`, see https://codex.wordpress.org/I18n_for_WordPress_Developers)
* [ ] My code is tested.
* [ ] My code follows the Event Espresso code style.
* [ ] My code has proper inline documentation.
* [ ] My code accounts for when the site is in Maintenance Mode (MM2 especially disallows usage of models)

### CODE REVIEW

**Comment by mnelson4** [comment]
## Several items I'd like Feedback on

* [ ] would you expect `POST /questions/1/question_groups/2` to add a new row to `question_group_question` every time a request is sent? (does it depend on the value of `QGQ_order`?)
In the case of `question_group_question`, we might prefer to avoid having duplicate rows (because you can't do that via the web UI). `registration_payment` I'm not sure if we want to allow multiple rows for how much each payment counts towards a registration's amount due...
I'm leaning towards allowing duplicate rows in join tables (because it's theoretically possible some models might want that, eg `extra_join` could); but I also realize that if someone really wants to do that, they can always directly create rows in the table, they don't need to use these convenience relation-editing endpoints.

* [ ] and would you expect `DELETE /questions/1/question_groups/2` to remove ALL entries in the join table or just one? I'm also slightly leaning towards deleting ALL join-table entries when you request to remove a relation (otherwise we'll need a way to indicate "request successfully deleted a join table row, but others exist" I think)
* [ ] should I add entries to `_links` for these new routes? I don't really think so, as those don't specify which HTTP methods to send with the requests
* [ ] Should URLs have singular entity names in them? eg is datetimes/1/event/2 better or worse than  datetimes/1/events/2? Currently it's `datetimes/1/event/2`, as the routes are the same as the relation reading endpoints (eg `GET datetimes/1/event` gets the datetime's single event), it just also accepting another ID.

**Comment by nerrad** [comment]
> would you expect POST /questions/1/question_groups/2 to add a new row to question_group_question every time a request is sent? (does it depend on the value of QGQ_order?)

Behaviour wise, what benefit is there to having multiple rows for that relation?  Although it may be possible I think functionally wise, any code implementing this data would have the expectation there is only one record per question_group/question record right? So imo it should only allow one record per relation.

The other problem with allowing multiple records to get created is the side effect of how the join table data is represented in the `_calculated_fields` for a `GET` request on the relation endpoint (eg. `questions/1/question_groups`), currently it's just a singular representation of the join table data so that would be broken.

> and would you expect DELETE /questions/1/question_groups/2 to remove ALL entries in the join table or just one?

Again, functionally speaking, what action performed by client code would have the expectation there'd be multiple relation rows for each question group -> question record?  In other words, should we even be getting into the state of having multiple records for these relations in the table?  So, on the off chance there are somehow, I'm fine with DELETE removing all entries in the table for that particular relation.

> should I add entries to _links for these new routes? I don't really think so, as those don't specify which HTTP methods to send with the requests

You could have a `self` link for each entity returned in the response?  That way client code knows that it can retrieve that entity schema etc via the `self` link.

> Should URLs have singular entity names in them? eg is datetimes/1/event/2 better or worse than datetimes/1/events/2?

Whatever you do, please be consistent for ALL routes (i.e. not singular for some routes because there is only the potential for a single relation).  Having a mix increases the logic client code needs to have for programatically generating the right endpoint.

**Comment by mnelson4** [comment]
I want to be thrifty with adding more routes to the WP API index, as we've already got so many that we've had errors storing them in the WP options table (and generating them via the models on every REST API request is a significant time sink). So I'd rather stick with the existing routes, which are sometimes singular (eg `datetimes/1/event/2`).
The entities returned from a request like `POST datetimes/1/event/2` DO have `_links` in them, including a `self` one. So I think we're good there.

kk, so to-do
* [ ] when adding relations, if extra parameters don't match an existing row, update it
* [ ] when removing relations, remove ALL
* [ ] don't support extra query parameters when removing relations (REST API assumes only one per join entry)
* [ ] update documentation

**Comment by mnelson4** [comment]
Once again, @nerrad could you please review that the REST API interface meets your needs, afterwards I'll ask Brent to review the code internals.

**Comment by mnelson4** [comment]
Ok @tn3rb could you give the code a review?

File: docs/C--REST-API/ee4-rest-api-editing-relations.md [hunk 1]
```diff
@@ -0,0 +1,352 @@
...
+            "_links": {
+                ...
+            },
+            "_calculated_fields": {}
+        }
+    }
+}
+```
+
+to indicate the relation was successful removed. In this case, the entry in `join.question_group_question` is actually the
```
[the remark this answers is not in the document]
Comment (mnelson4) [reply docs/C--REST-API/ee4-rest-api-editing-relations.md]: :+1: thanks

File: docs/C--REST-API/ee4-rest-api-editing-relations.md [hunk 2]
```diff
@@ -0,0 +1,352 @@
...
+**previous** question-group-question entry that was removed during this request.
+If no question-group-question were removed during the request, `join.question_group_question` will be null.
```
[the remark this answers is not in the document]
Comment (mnelson4) [reply docs/C--REST-API/ee4-rest-api-editing-relations.md]: :+1: thanks again

**Comment by nerrad** [comment]
I just want to highlight that this does block #790 from being complete as I need to wire things up there to work with the new endpoints and test appropriately.  While I can merge this branch in, I'd be more comfortable with landing this in master first so I can avoid potential merge conflicts and this is baked a bit in master before releasing #790.

**Comment by mnelson4** [comment]
I had heard we were asking Josh to do more of the triaging now, so passing to you Josh instead for assigning to someone for testing. (Please let me know if I should handle differently.)

**Comment by joshfeck** [comment]
tested & ready for merge, pending unit test fixes.

**Comment by mnelson4** [comment]
It seems there was some change to Travis that's causing that (I think Brent was basically reporting the same issue here: https://eventespresso.slack.com/archives/C04J554E8/p1545942570001900).
So this is probably ok, but we should probably wait until this issue with Travis is resolved.
I think Darren has the best idea of how to fix the Travis issue. Darren, could you either merge this, or update it, when we're working ok with Travis again?

**Pull request merged**